src/structured_config.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Structured configuration for the Superset charm."""
+import logging
+from enum import Enum
+from typing import Optional
+
+from charms.data_platform_libs.v0.data_models import BaseConfigModel
+from pydantic import validator
+
+logger = logging.getLogger(__name__)
+
+
+class BaseEnumStr(str, Enum):
+    """Base class for string enum."""
+
+    def __str__(self) -> str:
+        """Return the value as a string.
+
+        Returns:
+            string of config value
+        """
+        return str(self.value)
+
+
+class LogLevelType(str, Enum):
+    """Enum for the `log-level` field."""
+
+    INFO = "INFO"
+    DEBUG = "DEBUG"
+    WARNING = "WARNING"
+    ERROR = "ERROR"
+    FATAL = "FATAL"
+
+
+class StorageType(str, Enum):
+    """Enum for the `storage-type` field."""
+
+    minio = "minio"
+
+
+class CharmConfig(BaseConfigModel):
+    """Manager for the structured configuration."""
+
+    log_level: LogLevelType
+    temporal_host: str
+    storage_type: StorageType
+    storage_bucket_log: str
+    storage_bucket_state: str
+    storage_bucket_activity_payload: str
+    storage_bucket_workload_output: str
+    pod_running_ttl_minutes: int
+    pod_successful_ttl_minutes: int
+    pod_unsuccessful_ttl_minutes: int
+
+    @validator("*", pre=True)
+    @classmethod
+    def blank_string(cls, value):
+        """Check for empty strings.
+
+        Args:
+            value: configuration value
+
+        Returns:
+            None in place of empty string or value
+        """
+        if value == "":
+            return None
```

Comment:
nitpick: If I understand correctly, this runs for every field in the config, but type hints for those do not indicate `Optional` or `None`.

Reply:
This is because all of the listed configs are required. For ones that are optional, the type hint would be `Optional[str]`, in which case the charm will not go into an error state if the config value is `""`.

Reply:
Do you know why do we need to convert things to None though? Mert has a point there. If a variable is of type str and not Optional[str], I would assume it will never None, but here we end up with a None anyway. I think i have the same code in the governor, but not sure why we need this validator.

Reply:
@AmberCharitos might be able to correct me if i'm wrong, but the idea is that if a user sets a required string config value to `""`, the structured config will accept it as a valid value. Setting an empty string to `None` indicates to the structured config that this required string config does not have a valid value, hence sending the charm into an error state.

Reply:
For a required config value, `""` is not a valid value. I believe without this validator, required config values with a value of `""` will pass the check when they shouldn't, hence why they are converted to `None`.

Reply:
If `""` is not a valid value, why not raise a `ValueError`?

Reply:
I understand. As I see it, there are three scenarios here:

1. The config value `x` has a type hint of `str`, and the user sets the config value to `""`. We would want this to not pass the validation. Without this validator, it would pass the validation because `""` is a valid string value, while `None` is not.
2. The config value `x` has a type hint of `str`, and the user sets the config value to `"abc"`. This would pass the validator.
3. The config value `x` has a type hint of `Optional[str]`, and the user sets the config value to `""`. This would pass the validation, even though we are returning `None`, because it passes the type check.

There is a chance I remove the structured config depending on how it goes further into the development process.

Reply:
Question is, when you pass `""` to this code, do those config values ever get `None` or do they error out?

Reply:
If I set a config value with type hint `str` to `""` in the config, the charm errors out, because this method sets their value to `None` which is not an allowed `str` value.

Reply:
I understand now, thanks.